lib/rbac/filterer.rb

```diff
@@ -221,7 +221,9 @@ def search(options = {})
       end
 
       if search_filter && targets && (!exp_attrs || !exp_attrs[:supported_by_sql])
-        rejects     = targets.reject { |obj| matches_search_filters?(obj, search_filter, tz) }
+        rejects = TimeProfile.with_cached_time_profile_ids do
```

Comment:
This is tough to explain since there is a lot of meta-programming in the way for some of the columns I am referring to that use these, but I will see what I can do (also, I don't fully understand the usecase of `TimeProfiles` myself, just how they are queried against really).

So `TimeProfile::with_cached_time_profile_ids` doesn't grab all of the `TimeProfiles`, but one for each region based off of the default one, so really this part taking this portion of [this method from `Metric::Common`](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/metric/common.rb#L171-L173):

```ruby
def with_time_profile_or_tz(time_profile_or_tz = nil)
  if (time_profile = TimeProfile.default_time_profile(time_profile_or_tz))
    tp_ids = time_profile.profile_for_each_region
```

and saving the ids from at.  This does two calls against the `time_profiles` database table, but each throws away some records.

One to get the time profile (part of the `if` in the first line) by calling [`TimeProfile::default_time_profile`](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L186-L190).  This originally selects all [profiles where `rollup_daily_metrics` is `true`](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L13) in the database field, and [throws away ones](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L18-L20) that don't include all of the days of the week and all of the hours of the day (has to be [done in ruby](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L92-L94) because it compares the [days](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L72-L74) and [hours](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L82-L84) parts of the `profile` column, which is serialized yml), and only selects the first one that matches the passed in time zone or the default (also has to be [done in ruby](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L68-L70) since the [`tz`](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L59-L61) is a part of the profile field as well).

The second is part of the next line, or [`TimeProfile#profile_each_region` instance method](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L128-L139).  This again [starts as a sql select](https://github.com/ManageIQ/manageiq/blob/de74a31/app/models/time_profile.rb#L13) on the `rollup_daily_metrics` column, and then does an array `.select` to do an equivalence check against the returned profile columns against the original (`.select { |p| p.profile == profile }`), turning it into a hash by `region_id` (`.group_by(&:region_id)`), selecting the values from the generated hash, and then only returning the first `TimeProfile` for each region id returned (`map(&:first)`).

In the DB I was using, there were 5 regions in total, and 7 `TimeProfiles`, so only 5 of those time profiles were ever used when using this new method. or the original it is based off of.

Now, this was important because I didn't want to bring back extra data when doing queries for the actual `virtual_columns` that make use of these `time_profile` ids.  So for example, when calling `cpu_usagemhz_rate_average_max_over_time_period`:

* defined [here](https://github.com/ManageIQ/manageiq/blob/de74a31b/app/models/vm_or_template/right_sizing.rb#L139-L142)
* calls [`VimPerformanceAnalysis.find_perf_for_time_period`](https://github.com/ManageIQ/manageiq/blob/de74a31b/app/models/vim_performance_analysis.rb#L428-L441)
* which calls [`Metric::Helper.find_for_interval_name`](https://github.com/ManageIQ/manageiq/blob/de74a31b/app/models/metric/helper.rb#L10-L14)
* which calls `rel.with_time_profile_or_tz`, which is [part of `Metric::Common`](https://github.com/ManageIQ/manageiq/blob/de74a31b/app/models/metric/common.rb#L170-L178) and is included in both `Metric` and `Metric::Rollup` (this is what I updated the caching with in the PR).

This is fine for all of these `virtual_columns`, and those shouldn't be defining custom `TimeProfile`s (as far as I can tell), but I didn't want to bring all of them back because we were defining `tp_ids` for a specific set of them, and I wanted to make sure we weren't caching them longer than we needed to, (though, I just overwrite this method, so I guess honestly it probably is fine, since it will only ever return the "default ones" anyway now that I think about it).

Hopefully that made some semblance of sense... I spent a good chunk of time figuring this out myself and the call stack to get here, and am still trying to get it fully figured out, but let me know if anything I just explained doesn't make sense or is just plain wrong (definitely possible...).